### PR TITLE
Previous elections pages

### PR DESCRIPTION
**Pull request by tomrndom**
* Reuses template for previous elections pages
* Generate pages in build time with data from API

ref: https://tipit.avaza.com/project/view/188886#!tab=task-pane&task=3400289

Signed-off-by: Tomás Castillo <tcastilloboireau@gmail.com>

### CODE REVIEW

File: gatsby-node.js [hunk 1]
```diff
@@ -6,6 +6,14 @@ const axios = require('axios')
 const {createFilePath} = require('gatsby-source-filesystem')
 const {fmImagesToRelative} = require('gatsby-remark-relative-images')
 const {ClientCredentials} = require('simple-oauth2');
+const yaml = require("yaml")
+const moment = require("moment-timezone");
+const prevElectionsBasePath = 'src/pages/election/previous-elections';
+const electionsSinceYear = 2023;
```
Comment (smarcet) [inline gatsby-node.js:12]: this should be a env variable GATSBY_ELECTION_SINCE_YEAR ( default 2023)

File: gatsby-node.js [hunk 2]
```diff
@@ -6,6 +6,14 @@ const axios = require('axios')
...
+const currentYear = new Date().getFullYear();
+const minimunElectionsToShow = 2;
```
Comment (smarcet) [inline gatsby-node.js:14]: this should be a env variable GATSBY_ELECTION_TO_SHOW ( default 2)

File: gatsby-node.js [hunk 3]
```diff
@@ -169,9 +226,166 @@ exports.onPreBootstrap = async () => {
...
+            console.log(err);
+          }
+        });
+
+      if(!fs.existsSync(`${prevElectionsBasePath}/candidates/gold/${election.id}_gold_candidates.md`))
+        fs.writeFileSync(`${prevElectionsBasePath}/candidates/gold/${election.id}_gold_candidates.md`, `---\n${yaml.stringify({
+            templateKey: 'election-gold-candidates-page-previous',
+            electionYear:electionYear,
+            electionId:election.id,
+            title:election.name + ' Gold Candidates',
```
Comment (smarcet) [inline gatsby-node.js:324]: use string interpolation

File: gatsby-node.js [hunk 4]
```diff
@@ -239,16 +453,83 @@ exports.createSchemaCustomization = ({actions}) => {
...
+      nominationOpens: Int
+      nominationCloses: Int
+      nominationApplicationDeadline: Int
+    }
     `
   createTypes(typeDefs)
 }
 
-exports.createPages = ({actions, graphql}) => {
+exports.createPages = async ({actions, graphql}) => {
```
Comment (smarcet) [inline gatsby-node.js:467]: remove async, its not used

File: gatsby-node.js [hunk 5]
```diff
@@ -239,16 +453,83 @@ exports.createSchemaCustomization = ({actions}) => {
...
   const {createPage} = actions
 
+  await graphql(`
```
Comment (smarcet) [inline gatsby-node.js:470]: remove await is not used

File: gatsby-node.js [hunk 6]
```diff
@@ -277,12 +559,13 @@ exports.createPages = ({actions, graphql}) => {
...
     }
 
-    const pages = result.data.allMarkdownRemark.edges
+    const pages = result.data.allMarkdownRemark.edges;
 
     pages.forEach(edge => {
       if (edge.node.frontmatter.templateKey) {
         const id = edge.node.id
         const SEO = edge.node.frontmatter.seo ? edge.node.frontmatter.seo : null;
+        const electionId = edge.node.frontmatter.electionId ? `${edge.node.frontmatter.electionId}` : null
```
Comment (smarcet) [inline gatsby-node.js:568]: this , i do think, is not needed anymore

File: gatsby-node.js [hunk 7]
```diff
@@ -293,6 +576,7 @@ exports.createPages = ({actions, graphql}) => {
           // additional data can be passed via context
           context: {
             id,
+            electionId
```
Comment (smarcet) [inline gatsby-node.js:579]: this , i do think, is not needed anymore

File: static/admin/config.yml [hunk 8]
```diff
@@ -1203,3 +1203,54 @@ collections:
...
           - { label: "Sub Title", name: subTitle, widget: string }
           - { label: "Body", name: body, widget: markdown }
+  - name: "previous-election-pages"
+    label: "Previous Election Main Pages"
+    folder: "/src/pages/election/previous-elections"
+    editor:
+      preview: false
+    identifier_field: title
+    fields:
+      - { label: "Election Id", name: electionId, widget: string, required: true}
```
Comment (smarcet) [inline static/admin/config.yml:1213]: make readonly

File: static/admin/config.yml [hunk 9]
```diff
@@ -1203,3 +1203,54 @@ collections:
...
+      - { label: "Title", name: title, widget: string, required: true }
+      - { label: "Body", name: body, widget: markdown, required: true }
+  - name: "previous-election-page-gold-candidates"
+    label: "Previous Election Gold Candidates Pages"
+    folder: "/src/pages/election/previous-elections/candidates/gold"
+    editor:
+      preview: false
+    identifier_field: title
+    fields:
+      - { label: "Election Id", name: electionId, widget: string, required: true }
```
Comment (smarcet) [inline static/admin/config.yml:1247]: readonly

File: src/pages/election/index.md [hunk 10]
```diff
@@ -7,20 +7,20 @@ seo:
...
+    link: /election/individual-director-election
   - text: SEE THE CANDIDATES
-    link: /election/2023-individual-director-election/candidates
+    link: /election/individual-director-election/candidates
   - text: NOMINATE A MEMBER
     link: /a/community/members
   - text: BE A CANDIDATE
     link: /profile
   - text: GOLD MEMBER ELECTION CANDIDATES
-    link: /election/2023-individual-director-election/candidates/gold
```
Comment (smarcet) [inline src/pages/election/index.md:23]: rollback this file see https://github.com/OpenStackweb/osf-website/pull/517

File: src/pages/election/candidates/index.md [hunk 11]
```diff
@@ -7,19 +7,19 @@ seo:
   image: /img/OpenInfra-icon-white.jpg
   title: 2023 Board Elections - Candidates List
   twitterUsername: "@OpenInfraDev"
-  url: https://openinfra.dev/election/2023-individual-director-election/candidates
+  url: https://openinfra.dev/election/individual-director-election/candidates
```
Comment (smarcet) [inline src/pages/election/candidates/index.md:10]: rollback this file
see https://github.com/OpenStackweb/osf-website/pull/517

File: src/pages/election/candidates/gold/index.md [hunk 12]
```diff
@@ -7,19 +7,19 @@ seo:
...
   title: 2023 Board Elections - Gold Candidates List
   twitterUsername: "@OpenInfraDev"
-  url: https://openinfra.dev/election/2023-individual-director-election/candidates/gold
+  url: https://openinfra.dev/election/individual-director-election/candidates/gold
 title: 2023 Board Elections - Gold Candidates List
 menu:
   - text: ELECTION DETAILS
-    link: /election/2023-individual-director-election
+    link: /election/individual-director-election
   - text: SEE THE CANDIDATES
```
Comment (smarcet) [inline src/pages/election/candidates/gold/index.md:15]: rollback this file 
see https://github.com/OpenStackweb/osf-website/pull/517

File: gatsby-node.js [hunk 13]
```diff
@@ -6,6 +6,14 @@ const axios = require('axios')
 const {createFilePath} = require('gatsby-source-filesystem')
 const {fmImagesToRelative} = require('gatsby-remark-relative-images')
 const {ClientCredentials} = require('simple-oauth2');
+const yaml = require("yaml")
+const moment = require("moment-timezone");
+const prevElectionsBasePath = 'src/pages/election/previous-elections';
+const currentYear = new Date().getFullYear();
+const electionsSinceYear = process.env.GATSBY_ELECTION_SINCE_YEAR;
```
Comment (smarcet) [inline gatsby-node.js:13]: need default value ( 2013) is not defined

**Review by smarcet — APPROVED** [review]
LGTM
before merge we need to change the branch on CMS config and remove the extra config from .toml file

**Comment by iamweswilson** [comment]
I updated the CMS and that worked well. I'm not sure if I missed some areas in the CMS, but ran into an inconsistency when clicking through the preview. As you can see in this video, going to "Gold Member Elections" takes you to the right url, but the wrong title. Clicking "Election Details" from that page takes you back to 2023 elections. 

recording


<details><summary>Browser metadata</summary>





```
Path:      /election/2023-individual-director-election
Browser:   Chrome 118.0.0.0 on Mac OS 10.15.7
Viewport:  1404 x 1072 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-518--openinfra-dev.netlify.app%2Felection%2F2023-individual-director-election&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=118.0&resolution=1404x1072&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-518--openinfra-dev.netlify.app/election/2023-individual-director-election) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=653ffbf1fe19000008f8db29&commentID=654031e94f3afc0a98802189&resolution=resolved)

**Pull request merged**